Windows now passing CI.
Linux i686 is passing CI.
Linux x86_64 is failing early in the build with [`make[1]: *** [julia-deps] Error 2`](https://travis-ci.org/JuliaLang/julia/jobs/103527838#L3455)

--- a/base/filesystem.jl
+++ b/base/filesystem.jl
@@ -158,19 +158,26 @@ function read!(f::File, a::Vector{UInt8}, nel=length(a))
     end
     ret = ccall(:jl_fs_read, Int32, (Int32, Ptr{Void}, Csize_t),
                 f.handle, a, nel)
+    if ret < nel
+        throw(EOFError())
+    end
     uv_error("read",ret)
     return a
 end
 
 nb_available(f::File) = filesize(f) - position(f)
 
+eof(f::File) = nb_available(f) == 0
+
 function readbytes!(f::File, b::Array{UInt8}, nb=length(b))
     nr = min(nb, nb_available(f))
     if length(b) < nr
         resize!(b, nr)
     end
-    read!(f, b, nr)
-    return nr
+    ret = ccall(:jl_fs_read, Int32, (Int32, Ptr{Void}, Csize_t),
+                f.handle, b, nr)
+    uv_error("read",ret)
+    return ret
 end
 read(io::File) = read!(io, Array(UInt8, nb_available(io)))
 read(io::File, nb::Integer) = read!(io, Array(UInt8, min(nb, nb_available(io))))

--- a/base/filesystem.jl
+++ b/base/filesystem.jl
@@ -40,6 +40,7 @@ export File,
 
 import Base: uvtype, uvhandle, eventloop, fd, position, stat, close,
             write, read, read!, isopen, show,
+            seek, seekend, skip, eof,
             check_open, _sizeof_uv_fs, uv_error, UVError
 
 include("path.jl")
@@ -185,6 +186,24 @@ read(io::File, nb::Integer) = read!(io, Array(UInt8, min(nb, nb_available(io))))
 const SEEK_SET = Int32(0)
 const SEEK_CUR = Int32(1)
 const SEEK_END = Int32(2)
+
+function seek(f::File, n::Integer)
+    ret = ccall(:jl_lseek, Int64, (Int32, Int64, Int32), f.handle, n, SEEK_SET)
+    systemerror("seek", ret == -1)
+    return f
+end
+
+function seekend(f::File)
+    ret = ccall(:jl_lseek, Int64, (Int32, Int64, Int32), f.handle, 0, SEEK_END)
+    systemerror("seekend", ret == -1)
+    return f
+end
+
+function skip(f::File, n::Integer)
+    ret = ccall(:jl_lseek, Int64, (Int32, Int64, Int32), f.handle, n, SEEK_CUR)
+    systemerror("skip", ret == -1)
+    return f
+end
 
 function position(f::File)
     check_open(f)

--- a/base/io.jl
+++ b/base/io.jl
@@ -348,7 +348,10 @@ type EachLine
     EachLine(stream, ondone) = new(stream, ondone)
 end
 eachline(stream::IO) = EachLine(stream)
-eachline(filename::AbstractString) = EachLine(open(filename), close)
+function eachline(filename::AbstractString)
+    s = open(filename)
+    EachLine(s, ()->close(s))
+end
 
 start(itr::EachLine) = nothing
 function done(itr::EachLine, nada)

--- a/src/builtins.c
+++ b/src/builtins.c
@@ -182,6 +182,14 @@ JL_DLLEXPORT void JL_NORETURN jl_bounds_error_ints(jl_value_t *v, size_t *idxs,
     jl_throw(jl_new_struct((jl_datatype_t*)jl_boundserror_type, v, t));
 }
 
+JL_DLLEXPORT void JL_NORETURN jl_eof_error(void)
+{
+    jl_datatype_t *eof_error =
+        (jl_datatype_t*)jl_get_global(jl_base_module, jl_symbol("EOFError"));
+    assert(eof_error != NULL);
+    jl_exceptionf(eof_error, "");
+}
+
 JL_CALLABLE(jl_f_throw)
 {
     JL_NARGS(throw, 1, 1);

--- a/src/jl_uv.c
+++ b/src/jl_uv.c
@@ -339,9 +339,12 @@ JL_DLLEXPORT int jl_fs_read_byte(int handle)
     buf[0].len = 1;
     int ret = uv_fs_read(jl_io_loop, &req, handle, buf, 1, -1, NULL);
     uv_fs_req_cleanup(&req);
-    if (ret == -1)
-        return ret;
-    return (int)c;
+    switch (ret) {
+        case -1: return ret;
+        case  0: jl_eof_error();
+        case  1: return (int)c;
+        default: assert(0);
+    }
 }
 
 JL_DLLEXPORT int jl_fs_close(int handle)

--- a/src/julia.h
+++ b/src/julia.h
@@ -1250,6 +1250,7 @@ JL_DLLEXPORT void JL_NORETURN jl_bounds_error_tuple_int(jl_value_t **v,
                                                         size_t nv, size_t i);
 JL_DLLEXPORT void JL_NORETURN jl_bounds_error_unboxed_int(void *v, jl_value_t *vt, size_t i);
 JL_DLLEXPORT void JL_NORETURN jl_bounds_error_ints(jl_value_t *v, size_t *idxs, size_t nidxs);
+JL_DLLEXPORT void JL_NORETURN jl_eof_error(void);
 JL_DLLEXPORT jl_value_t *jl_exception_occurred(void);
 JL_DLLEXPORT void jl_exception_clear(void);
 

--- a/src/sys.c
+++ b/src/sys.c
@@ -308,13 +308,6 @@ JL_DLLEXPORT jl_value_t *jl_readuntil(ios_t *s, uint8_t delim)
     return (jl_value_t*)a;
 }
 
-static void JL_NORETURN throw_eof_error(void)
-{
-    jl_datatype_t *eof_error = (jl_datatype_t*)jl_get_global(jl_base_module, jl_symbol("EOFError"));
-    assert(eof_error != NULL);
-    jl_exceptionf(eof_error, "");
-}
-
 JL_DLLEXPORT uint64_t jl_ios_get_nbyte_int(ios_t *s, const size_t n)
 {
     assert(n <= 8);
@@ -323,7 +316,7 @@ JL_DLLEXPORT uint64_t jl_ios_get_nbyte_int(ios_t *s, const size_t n)
         space = (size_t)(s->size - s->bpos);
         ret = ios_readprep(s, n);
         if (space == ret && ret < n)
-            throw_eof_error();
+            jl_eof_error();
     } while(ret < n);
     uint64_t x = 0;
     uint8_t *buf = (uint8_t*)&s->buf[s->bpos];

--- a/test/choosetests.jl
+++ b/test/choosetests.jl
@@ -23,7 +23,7 @@ function choosetests(choices = [])
         "bitarray", "copy", "math", "fastmath", "functional",
         "operators", "path", "ccall", "parse", "loading",
         "bigint", "sorting", "statistics", "spawn", "backtrace",
-        "priorityqueue", "file", "mmap", "version", "resolve",
+        "priorityqueue", "file", "read", "mmap", "version", "resolve",
         "pollfd", "mpfr", "broadcast", "complex", "socket",
         "floatapprox", "datafmt", "reflection", "regex", "float16",
         "combinatorics", "sysinfo", "rounding", "ranges", "mod2pi",

--- a/test/file.jl
+++ b/test/file.jl
@@ -1115,3 +1115,36 @@ end
 @test copy(DevNull) === DevNull
 @test eof(DevNull)
 @test print(DevNull, "go to /dev/null") === nothing
+
+# Filesystem.File
+tmpdir = mktempdir() do d
+    f = joinpath(d, "test.txt")
+    open(io->write(io, "123"), f, "w")
+    f1 = open(f)
+    f2 = Base.Filesystem.open(f, Base.Filesystem.JL_O_RDONLY)
+    @test read(f1, UInt8) == read(f2, UInt8)
+    @test read(f1, UInt8) == read(f2, UInt8)
+    @test read(f1, UInt8) == read(f2, UInt8)
+    @test_throws EOFError read(f1, UInt8)
+    @test_throws EOFError read(f2, UInt8)
+    close(f1)
+    close(f2)
+
+    a = UInt8[0,0,0]
+    f1 = open(f)
+    f2 = Base.Filesystem.open(f, Base.Filesystem.JL_O_RDONLY)
+    @test read!(f1, a) == read!(f2, a)
+    @test_throws EOFError read!(f1, a)
+    @test_throws EOFError read!(f2, a)
+    close(f1)
+    close(f2)
+
+    a = UInt8[0,0,0,0]
+    f1 = open(f)
+    f2 = Base.Filesystem.open(f, Base.Filesystem.JL_O_RDONLY)
+    @test_throws EOFError read!(f1, a)
+    @test_throws EOFError read!(f2, a)
+    close(f1)
+    close(f2)
+end
+

--- a/test/read.jl
+++ b/test/read.jl
@@ -12,7 +12,7 @@ l = Vector{Tuple{AbstractString,Function}}()
 
 # File
 io = (text) -> begin
-    open(io-> write(io, text), filename, "w")
+    write(filename, text)
     Base.Filesystem.open(filename, Base.Filesystem.JL_O_RDONLY)
 end
 s = io(text)
@@ -24,7 +24,7 @@ push!(l, ("File", io))
 
 # IOStream
 io = (text) -> begin
-    open(io-> write(io, text), filename, "w")
+    write(filename, text)
     open(filename)
 end
 s = io(text)
@@ -100,7 +100,7 @@ push!(l, ("PipeEndpoint", io))
 
 # Pipe
 io = (text) -> begin
-    open(io->write(io, text), filename, "w")
+    write(filename, text)
     open(`$(@windows ? "type" : "cat") $filename`)[1]
 #    Was open(`echo -n $text`)[1]
 #    See https://github.com/JuliaLang/julia/issues/14747
@@ -128,18 +128,20 @@ end
 
 verbose = false
 
-cleanup()
+                                                                       cleanup()
 
 for (name, f) in l
 
     io = ()->(s=f(text); push!(open_streams, s); s)
 
+    write(filename, text)
+
     verbose && println("$name read...")
     @test read(io(), UInt8) == read(IOBuffer(text), UInt8)
-    @test read(io(), UInt8) == open(io->read(io, UInt8), filename)
+    @test read(io(), UInt8) == read(filename, UInt8)
     @test read(io(), Int) == read(IOBuffer(text), Int)
-    @test read(io(), Int) == open(io->read(io,Int),filename)
-    cleanup()
+    @test read(io(), Int) == read(filename,Int)
+                                                                       cleanup()
     s1 = io()
     s2 = IOBuffer(text)
     @test read(s1, UInt32, 2) == read(s2, UInt32, 2)
@@ -176,13 +178,19 @@ for (name, f) in l
         UTF8String(['A' + i % 52 for i in 1:(    Base.SZ_UNBUFFERED_IO +1)])
     ]
 
-        verbose && println("$name readall...")
-        @test readall(io()) == text
-        cleanup()
+        write(filename, text)
 
-        verbose && println("$name readbytes...")
-        @test readbytes(io()) == Vector{UInt8}(text)
-        cleanup()
+        verbose && println("$name readstring...")
+        @test readstring(io()) == text
+                                                                       cleanup()
+        @test readstring(io()) == readstring(filename)
+                                                                       cleanup()
+
+        verbose && println("$name read...")
+        @test read(io()) == Vector{UInt8}(text)
+                                                                       cleanup()
+        @test read(io()) == read(filename)
+                                                                       cleanup()
 
         verbose && println("$name readbytes!...")
         l = length(text)
@@ -198,7 +206,7 @@ for (name, f) in l
             @test a1[1:n1] == a2[1:n2]
             @test n <= length(text) || eof(s1)
             @test n <= length(text) || eof(s2)
-            cleanup()
+                                                                       cleanup()
         end
 
         verbose && println("$name read!...")
@@ -206,41 +214,58 @@ for (name, f) in l
         for n = [1, 2, l-2, l-1, l]
             @test read!(io(), Vector{UInt8}(n)) ==
                   read!(IOBuffer(text), Vector{UInt8}(n))
-            cleanup()
+                                                                       cleanup()
+            @test read!(io(), Vector{UInt8}(n)) ==
+                  read!(filename, Vector{UInt8}(n))
+                                                                       cleanup()
         end
         @test_throws EOFError read!(io(), Vector{UInt8}(length(text)+1))
 
-        cleanup()
+                                                                       cleanup()
 
         verbose && println("$name readuntil...")
         @test readuntil(io(), '\n') == readuntil(IOBuffer(text),'\n')
-        cleanup()
+                                                                       cleanup()
+        @test readuntil(io(), '\n') == readuntil(filename,'\n')
+                                                                       cleanup()
         @test readuntil(io(), "\n") == readuntil(IOBuffer(text),"\n")
-        cleanup()
+                                                                       cleanup()
+        @test readuntil(io(), "\n") == readuntil(filename,"\n")
+                                                                       cleanup()
         @test readuntil(io(), ',') == readuntil(IOBuffer(text),',')
-        cleanup()
+                                                                       cleanup()
+        @test readuntil(io(), ',') == readuntil(filename,',')
+                                                                       cleanup()
 
         verbose && println("$name readline...")
         @test readline(io()) == readline(IOBuffer(text))
-        cleanup()
+                                                                       cleanup()
+        @test readline(io()) == readline(filename)
+                                                                       cleanup()
 
         verbose && println("$name readlines...")
         @test readlines(io()) == readlines(IOBuffer(text))
-        cleanup()
+                                                                       cleanup()
+        @test readlines(io()) == readlines(filename)
+                                                                       cleanup()
         @test collect(eachline(io())) == collect(eachline(IOBuffer(text)))
-        cleanup()
+                                                                       cleanup()
+        @test collect(eachline(io())) == collect(eachline(filename))
+                                                                       cleanup()
 
         verbose && println("$name countlines...")
         @test countlines(io()) == countlines(IOBuffer(text))
-        cleanup()
+                                                                       cleanup()
 
         verbose && println("$name readcsv...")
         @test readcsv(io()) == readcsv(IOBuffer(text))
-        cleanup()
+                                                                       cleanup()
+        @test readcsv(io()) == readcsv(filename)
+                                                                       cleanup()
     end
 
     text = old_text
-
+    write(filename, text)
 
     if !(typeof(io()) in [Base.PipeEndpoint, Pipe, TCPSocket])
 
@@ -250,19 +275,19 @@ for (name, f) in l
         verbose && println("$name seek...")
         for n = 0:length(text)-1
             @test readlines(seek(io(), n)) == readlines(seek(IOBuffer(text), n))
-            cleanup()
+                                                                       cleanup()
         end
         verbose && println("$name skip...")
         for n = 0:length(text)-1
             @test readlines(seek(io(), n)) == readlines(seek(IOBuffer(text), n))
             @test readlines(skip(io(), n)) == readlines(skip(IOBuffer(text), n))
-            cleanup()
+                                                                       cleanup()
         end
         verbose && println("$name seekend...")
-        @test readall(seekend(io())) == ""
+        @test readstring(seekend(io())) == ""
     end
 
-    cleanup()
+                                                                       cleanup()
 end
 
 end

--- a/test/read.jl
+++ b/test/read.jl
@@ -1,0 +1,177 @@
+mktempdir() do dir
+
+tasks = []
+
+# Create test file...
+filename = joinpath(dir, "file.txt")
+text = "C1,C2\n1,2\na,b\n"
+open(io-> write(io, text), filename, "w")
+
+# List of IO producers...
+l = Vector{Tuple{AbstractString,Function}}()
+
+
+# File
+io = ()->Base.Filesystem.open(filename, Base.Filesystem.JL_O_RDONLY)
+s = io()
+@test isa(s, IO)
+@test isa(s, Base.Filesystem.File)
+close(s)
+push!(l, ("File", io))
+
+
+# IOStream
+io = ()->open(filename)
+s = io()
+@test isa(s, IO)
+@test isa(s, IOStream)
+close(s)
+push!(l, ("IOStream", io))
+
+
+# IOBuffer
+io = ()->IOBuffer(text)
+s = io()
+@test isa(s, IO)
+@test isa(s, IOBuffer)
+close(s)
+push!(l, ("IOBuffer", io))
+
+@windows ? nothing : begin
+
+# PipeEndpoint
+socketname = joinpath(dir, "socket")
+io = ()-> begin
+    c = Base.Condition()
+    tsk = @async begin
+        con = listen(socketname)
+        Base.notify(c)
+        sock = accept(con)
+        write(sock,text)
+        close(con)
+        close(sock)
+    end
+    push!(tasks, tsk)
+    wait(c)
+    connect(socketname)
+end
+s = io()
+@test isa(s, IO)
+@test isa(s, Base.PipeEndpoint)
+close(s)
+for tsk in tasks
+    wait(tsk)
+end
+push!(l, ("PipeEndpoint", io))
+
+
+# Pipe
+io = () -> open(`echo -n $text`)[1]
+s = io()
+@test isa(s, IO)
+@test isa(s, Pipe)
+close(s)
+push!(l, ("Pipe", io))
+
+end
+
+open_streams = []
+
+for (name, f) in l
+
+    io = ()->(s=f(); push!(open_streams, s); s)
+
+    #println("$name readall...")
+    @test readall(io()) == text
+    @test readall(io()) == readall(filename)
+
+    #println("$name read...")
+    @test readbytes(io()) == Vector{UInt8}(text)
+    @test readbytes(io()) == open(readbytes,filename)
+    @test read(io(), UInt8) == read(IOBuffer(text), UInt8)
+    @test read(io(), UInt8) == open(io->read(io, UInt8), filename)
+    @test read(io(), Int) == read(IOBuffer(text), Int)
+    @test read(io(), Int) == open(io->read(io,Int),filename)
+    s1 = io()
+    s2 = IOBuffer(text)
+    @test read(s1, UInt32, 2) == read(s2, UInt32, 2)
+    @test !eof(s1)
+    @test read(s1, UInt8, 5) == read(s2, UInt8, 5)
+    @test !eof(s1)
+    @test read(s1, UInt8, 1) == read(s2, UInt8, 1)
+    @test eof(s1)
+    @test_throws EOFError read(s1, UInt8)
+    @test eof(s1)
+    close(s1)
+    close(s2)
+
+    #println("$name readuntil...")
+    @test readuntil(io(), '\n') == open(io->readuntil(io,'\n'),filename)
+    @test readuntil(io(), "\n") == open(io->readuntil(io,"\n"),filename)
+    @test readuntil(io(), ',') == open(io->readuntil(io,','),filename)
+
+    #println("$name eof...")
+    n = length(text) - 1
+    @test read!(io(), Vector{UInt8}(n)) ==
+          read!(IOBuffer(text), Vector{UInt8}(n))
+    @test (s = io(); read!(s, Vector{UInt8}(n)); !eof(s))
+    n = length(text)
+    @test read!(io(), Vector{UInt8}(n)) ==
+          read!(IOBuffer(text), Vector{UInt8}(n))
+    @test (s = io(); read!(s, Vector{UInt8}(n)); eof(s))
+    n = length(text) + 1
+    @test_throws EOFError read!(io(), Vector{UInt8}(n))
+    @test_throws EOFError read!(io(), Vector{UInt8}(n))
+
+    #println("$name read!...")
+    for n = 1:length(text)
+        @test read!(io(), Vector{UInt8}(n)) ==
+              read!(IOBuffer(text), Vector{UInt8}(n))
+        @test read!(io(), Vector{UInt8}(n)) ==
+              open(io->read!(io, Vector{UInt8}(n)), filename)
+    end
+    @test_throws EOFError read!(io(), Vector{UInt8}(length(text)+1))
+
+    #println("$name readline...")
+    @test readline(io()) == readline(IOBuffer(text))
+    @test readline(io()) == open(readline,filename)
+
+    #println("$name readlines...")
+    @test readlines(io()) == readlines(IOBuffer(text))
+    @test readlines(io()) == open(readlines,filename)
+    @test collect(eachline(io())) == collect(eachline(IOBuffer(text)))
+
+    #println("$name countlines...")
+    @test countlines(io()) == countlines(IOBuffer(text))
+
+    #println("$name readcsv...")
+    @test readcsv(io()) == readcsv(IOBuffer(text))
+
+    if !(typeof(io()) in [Base.PipeEndpoint, Pipe])
+
+        #println("$name position...")
+        @test (s = io(); read!(s, Vector{UInt8}(4)); position(s))  == 4
+
+        #println("$name seek...")
+        for n = 0:length(text)-1
+            @test readlines(seek(io(), n)) == readlines(seek(IOBuffer(text), n))
+        end
+        #println("$name skip...")
+        for n = 0:length(text)-1
+            @test readlines(seek(io(), n)) == readlines(seek(IOBuffer(text), n))
+            @test readlines(skip(io(), n)) == readlines(skip(IOBuffer(text), n))
+        end
+        #println("$name seekend...")
+        @test readall(seekend(io())) == ""
+    end
+end
+
+for s in open_streams
+    try close(s) end
+end
+
+for tsk in tasks
+    wait(tsk)
+end
+
+end

